{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6a1984f8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "FFFFFFFFF..........\n",
      "======================================================================\n",
      "FAIL: test_fail (__main__.SaveTest)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/var/folders/hd/643ldtb90kl47t0v8x569lbr1mrhfd/T/ipykernel_56739/151546010.py\", line 19, in test_fail\n",
      "    self.expect_table_row_count_to_equal(df, 20)\n",
      "  File \"/Users/sam.treweek/Desktop/great_assertions/ve/lib/python3.9/site-packages/great_assertions-0.0.64-py3.9.egg/great_assertions.py\", line 199, in expect_table_row_count_to_equal\n",
      "    raise self.failureException(msg)\n",
      "AssertionError: expected row count is 20 the actual was 1 : \n",
      "\n",
      "======================================================================\n",
      "FAIL: test_fail2 (__main__.SaveTest)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/var/folders/hd/643ldtb90kl47t0v8x569lbr1mrhfd/T/ipykernel_56739/151546010.py\", line 51, in test_fail2\n",
      "    self.expect_table_row_count_to_equal(df, 55)\n",
      "  File \"/Users/sam.treweek/Desktop/great_assertions/ve/lib/python3.9/site-packages/great_assertions-0.0.64-py3.9.egg/great_assertions.py\", line 199, in expect_table_row_count_to_equal\n",
      "    raise self.failureException(msg)\n",
      "AssertionError: expected row count is 55 the actual was 1 : \n",
      "\n",
      "======================================================================\n",
      "FAIL: test_fail3 (__main__.SaveTest)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/var/folders/hd/643ldtb90kl47t0v8x569lbr1mrhfd/T/ipykernel_56739/151546010.py\", line 59, in test_fail3\n",
      "    self.expect_table_row_count_to_equal(df, 99)\n",
      "  File \"/Users/sam.treweek/Desktop/great_assertions/ve/lib/python3.9/site-packages/great_assertions-0.0.64-py3.9.egg/great_assertions.py\", line 199, in expect_table_row_count_to_equal\n",
      "    raise self.failureException(msg)\n",
      "AssertionError: expected row count is 99 the actual was 1 : \n",
      "\n",
      "======================================================================\n",
      "FAIL: test_fail4 (__main__.SaveTest)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/var/folders/hd/643ldtb90kl47t0v8x569lbr1mrhfd/T/ipykernel_56739/151546010.py\", line 79, in test_fail4\n",
      "    self.expect_table_row_count_to_be_greater_than(df, 4)\n",
      "  File \"/Users/sam.treweek/Desktop/great_assertions/ve/lib/python3.9/site-packages/great_assertions-0.0.64-py3.9.egg/great_assertions.py\", line 265, in expect_table_row_count_to_be_greater_than\n",
      "    raise self.failureException(msg)\n",
      "AssertionError: expected row count of at least 4 but the actual was 3 : \n",
      "\n",
      "======================================================================\n",
      "FAIL: test_fail5 (__main__.SaveTest)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/var/folders/hd/643ldtb90kl47t0v8x569lbr1mrhfd/T/ipykernel_56739/151546010.py\", line 99, in test_fail5\n",
      "    self.expect_table_has_no_duplicate_rows(df)\n",
      "  File \"/Users/sam.treweek/Desktop/great_assertions/ve/lib/python3.9/site-packages/great_assertions-0.0.64-py3.9.egg/great_assertions.py\", line 292, in expect_table_has_no_duplicate_rows\n",
      "    raise self.failureException(msg)\n",
      "AssertionError: Table contains duplicate rows : \n",
      "\n",
      "======================================================================\n",
      "FAIL: test_fail6 (__main__.SaveTest)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/var/folders/hd/643ldtb90kl47t0v8x569lbr1mrhfd/T/ipykernel_56739/151546010.py\", line 114, in test_fail6\n",
      "    self.expect_column_values_to_be_between(df, \"col_1\", 101, 301)\n",
      "  File \"/Users/sam.treweek/Desktop/great_assertions/ve/lib/python3.9/site-packages/great_assertions-0.0.64-py3.9.egg/great_assertions.py\", line 422, in expect_column_values_to_be_between\n",
      "    raise self.failureException(msg)\n",
      "AssertionError: Min value provided (101) must be less than column col_1 value of 100 : \n",
      "\n",
      "======================================================================\n",
      "FAIL: test_fail7 (__main__.SaveTest)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/var/folders/hd/643ldtb90kl47t0v8x569lbr1mrhfd/T/ipykernel_56739/151546010.py\", line 126, in test_fail7\n",
      "    self.expect_column_values_to_match_regex(\n",
      "  File \"/Users/sam.treweek/Desktop/great_assertions/ve/lib/python3.9/site-packages/great_assertions-0.0.64-py3.9.egg/great_assertions.py\", line 464, in expect_column_values_to_match_regex\n",
      "    raise self.failureException(msg)\n",
      "AssertionError: Column col_1 did not match regular expression, found BA151 : \n",
      "\n",
      "======================================================================\n",
      "FAIL: test_fail8 (__main__.SaveTest)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/var/folders/hd/643ldtb90kl47t0v8x569lbr1mrhfd/T/ipykernel_56739/151546010.py\", line 170, in test_fail8\n",
      "    self.expect_date_range_to_be_less_than(df, \"col_1\", \"2019-05-13\")\n",
      "  File \"/Users/sam.treweek/Desktop/great_assertions/ve/lib/python3.9/site-packages/great_assertions-0.0.64-py3.9.egg/great_assertions.py\", line 649, in expect_date_range_to_be_less_than\n",
      "    raise self.failureException(msg)\n",
      "AssertionError: Column col_1 date is greater or equal than 2019-05-13 found 2019-05-13 : \n",
      "\n",
      "======================================================================\n",
      "FAIL: test_fail9 (__main__.SaveTest)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/var/folders/hd/643ldtb90kl47t0v8x569lbr1mrhfd/T/ipykernel_56739/151546010.py\", line 185, in test_fail9\n",
      "    self.expect_date_range_to_be_more_than(df, \"col_1\", \"2015-10-01\")\n",
      "  File \"/Users/sam.treweek/Desktop/great_assertions/ve/lib/python3.9/site-packages/great_assertions-0.0.64-py3.9.egg/great_assertions.py\", line 691, in expect_date_range_to_be_more_than\n",
      "    raise self.failureException(msg)\n",
      "AssertionError: Column col_1 is less or equal than 2015-10-01 found 2015-10-01 : \n",
      "\n",
      "----------------------------------------------------------------------\n",
      "Ran 19 tests in 2.224s\n",
      "\n",
      "FAILED (failures=9)\n"
     ]
    }
   ],
   "source": [
    "import unittest\n",
    "\n",
    "from great_assertions import GreatAssertionResult, GreatAssertions\n",
    "from pyspark.sql import SparkSession\n",
    "\n",
    "spark = SparkSession.builder.getOrCreate()\n",
    "\n",
    "\n",
    "class SaveTest(GreatAssertions):\n",
    "    def setUp(self):\n",
    "        self.spark = SparkSession.builder.getOrCreate()\n",
    "\n",
    "    def test_fail(self):\n",
    "        df = self.spark.createDataFrame(\n",
    "            [\n",
    "                {\"col_1\": 100, \"col_2\": 10},\n",
    "            ]\n",
    "        )\n",
    "        self.expect_table_row_count_to_equal(df, 20)\n",
    "\n",
    "    def test_pass1(self):\n",
    "        df = self.spark.createDataFrame(\n",
    "            [\n",
    "                {\"col_1\": 100, \"col_2\": 10},\n",
    "            ]\n",
    "        )\n",
    "        self.expect_table_row_count_to_equal(df, 1)\n",
    "\n",
    "    def test_pass2(self):\n",
    "        df = self.spark.createDataFrame(\n",
    "            [\n",
    "                {\"col_1\": 100, \"col_2\": 10},\n",
    "            ]\n",
    "        )\n",
    "        self.expect_table_row_count_to_equal(df, 1)\n",
    "\n",
    "    def test_pass3(self):\n",
    "        df = self.spark.createDataFrame(\n",
    "            [\n",
    "                {\"col_1\": 100, \"col_2\": 10},\n",
    "            ]\n",
    "        )\n",
    "        self.expect_table_row_count_to_equal(df, 1)\n",
    "\n",
    "    def test_fail2(self):\n",
    "        df = self.spark.createDataFrame(\n",
    "            [\n",
    "                {\"col_1\": 100, \"col_2\": 10},\n",
    "            ]\n",
    "        )\n",
    "        self.expect_table_row_count_to_equal(df, 55)\n",
    "\n",
    "    def test_fail3(self):\n",
    "        df = self.spark.createDataFrame(\n",
    "            [\n",
    "                {\"col_1\": 100, \"col_2\": 10},\n",
    "            ]\n",
    "        )\n",
    "        self.expect_table_row_count_to_equal(df, 99)\n",
    "\n",
    "    def test_pass3(self):\n",
    "        df = self.spark.createDataFrame(\n",
    "            [\n",
    "                {\"col_1\": 100, \"col_2\": 10},\n",
    "                {\"col_1\": 200, \"col_2\": 20},\n",
    "                {\"col_1\": 300, \"col_2\": 30},\n",
    "            ]\n",
    "        )\n",
    "        self.expect_table_row_count_to_be_greater_than(df, 2)\n",
    "\n",
    "    def test_fail4(self):\n",
    "        df = self.spark.createDataFrame(\n",
    "            [\n",
    "                {\"col_1\": 100, \"col_2\": 10},\n",
    "                {\"col_1\": 200, \"col_2\": 20},\n",
    "                {\"col_1\": 300, \"col_2\": 30},\n",
    "            ]\n",
    "        )\n",
    "        self.expect_table_row_count_to_be_greater_than(df, 4)\n",
    "\n",
    "    def test_pass4(self):\n",
    "        df = self.spark.createDataFrame(\n",
    "            [\n",
    "                {\"col_1\": 100, \"col_2\": 10},\n",
    "                {\"col_1\": 100, \"col_2\": 20},\n",
    "                {\"col_1\": 300, \"col_2\": 30},\n",
    "            ]\n",
    "        )\n",
    "        self.expect_table_has_no_duplicate_rows(df)\n",
    "\n",
    "    def test_fail5(self):\n",
    "        df = self.spark.createDataFrame(\n",
    "            [\n",
    "                {\"col_1\": 100, \"col_2\": 10},\n",
    "                {\"col_1\": 100, \"col_2\": 10},\n",
    "                {\"col_1\": 300, \"col_2\": 30},\n",
    "            ]\n",
    "        )\n",
    "        self.expect_table_has_no_duplicate_rows(df)\n",
    "\n",
    "    def test_pass5(self):\n",
    "        # int\n",
    "        df = self.spark.createDataFrame(\n",
    "            [{\"col_1\": 100}, {\"col_1\": 200}, {\"col_1\": 300}]\n",
    "        )\n",
    "        self.expect_column_values_to_be_between(\n",
    "            df, \"col_1\", min_value=99, max_value=301\n",
    "        )\n",
    "\n",
    "    def test_fail6(self):\n",
    "        df = self.spark.createDataFrame(\n",
    "            [{\"col_1\": 100}, {\"col_1\": 200}, {\"col_1\": 300}]\n",
    "        )\n",
    "        self.expect_column_values_to_be_between(df, \"col_1\", 101, 301)\n",
    "\n",
    "    def test_pass6(self):\n",
    "        df = self.spark.createDataFrame(\n",
    "            [{\"col_1\": \"BA2\"}, {\"col_1\": \"BA15\"}, {\"col_1\": \"SW1\"}]\n",
    "        )\n",
    "        self.expect_column_values_to_match_regex(df, \"col_1\", \"^[a-zA-Z]{2}[0-9]{1,2}$\")\n",
    "\n",
    "    def test_fail7(self):\n",
    "        df = self.spark.createDataFrame(\n",
    "            [{\"col_1\": \"BA2\"}, {\"col_1\": \"BA151\"}, {\"col_1\": \"AAA13\"}]\n",
    "        )\n",
    "        self.expect_column_values_to_match_regex(df, \"col_1\", \"^[a-zA-Z]{2}[0-9]{1,2}$\")\n",
    "\n",
    "    def test_pass7(self):\n",
    "        df_fruits = [\n",
    "            {\"col_1\": \"Apple\"},\n",
    "            {\"col_1\": \"Orange\"},\n",
    "            {\"col_1\": \"Cherry\"},\n",
    "            {\"col_1\": \"Apricot(Summer)\"},\n",
    "        ]\n",
    "        fruits = set((\"Apple\", \"Orange\", \"Pear\", \"Cherry\", \"Apricot(Summer)\"))\n",
    "        df = self.spark.createDataFrame(df_fruits)\n",
    "\n",
    "        self.expect_column_values_to_be_in_set(df, \"col_1\", fruits)\n",
    "\n",
    "    def test_pass8(self):\n",
    "        df = self.spark.createDataFrame(\n",
    "            [\n",
    "                {\"col_1\": \"BA2\", \"col_2\": 10, \"col_3\": 10.45},\n",
    "                {\"col_1\": \"BA15\", \"col_2\": 20, \"col_3\": 10.45},\n",
    "                {\"col_1\": \"SW1\", \"col_2\": 30, \"col_3\": 10.45},\n",
    "            ]\n",
    "        )\n",
    "        self.expect_column_values_to_be_of_type(df, \"col_1\", str)\n",
    "\n",
    "    def test_pass9(self):\n",
    "        df = self.spark.createDataFrame(\n",
    "            [\n",
    "                {\"col_1\": \"2019-05-13\"},\n",
    "                {\"col_1\": \"2018-12-12\"},\n",
    "                {\"col_1\": \"2015-10-01\"},\n",
    "            ]\n",
    "        )\n",
    "        self.expect_date_range_to_be_less_than(df, \"col_1\", \"2019-05-14\")\n",
    "\n",
    "    def test_fail8(self):\n",
    "        df = self.spark.createDataFrame(\n",
    "            [\n",
    "                {\"col_1\": \"2019-05-13\"},\n",
    "                {\"col_1\": \"2018-12-12\"},\n",
    "                {\"col_1\": \"2015-10-01\"},\n",
    "            ]\n",
    "        )\n",
    "        self.expect_date_range_to_be_less_than(df, \"col_1\", \"2019-05-13\")\n",
    "\n",
    "    def test_pass10(self):\n",
    "        df = self.spark.createDataFrame([{\"col_1\": \"\"}])\n",
    "\n",
    "        self.expect_date_range_to_be_more_than(df, \"col_1\", \"1899-12-31\")\n",
    "\n",
    "    def test_fail9(self):\n",
    "        df = self.spark.createDataFrame(\n",
    "            [\n",
    "                {\"col_1\": \"2019-05-13\"},\n",
    "                {\"col_1\": \"2018-12-12\"},\n",
    "                {\"col_1\": \"2015-10-01\"},\n",
    "            ]\n",
    "        )\n",
    "        self.expect_date_range_to_be_more_than(df, \"col_1\", \"2015-10-01\")\n",
    "\n",
    "\n",
    "suite = unittest.TestLoader().loadTestsFromTestCase(SaveTest)\n",
    "test_runner = unittest.runner.TextTestRunner(resultclass=GreatAssertionResult)\n",
    "result = test_runner.run(suite)\n",
    "result.save(format=\"pyspark\", spark=spark)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
